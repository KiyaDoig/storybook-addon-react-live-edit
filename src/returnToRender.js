// @flow
import { className } from './constants';

export default (code: string): string => `const thunk = (render) => {
${code}
};
    
    let renderWasCalled = false;
    
    const renderWithCheck = (element) => {
        renderWasCalled = true;
        render(element);
    };
    
    const result = thunk(renderWithCheck);
    
    if (!renderWasCalled) {
        if (React.isValidElement(result)) {
            render(result);
        } else {
            render(<div className="${className.error}">Code should return a valid react element or call 'render(element)'!</div>);
        }
    }
`;

const safeSubtractFromString = (a: string, b: number): string => String(Math.max(Number(a) - b, 0));

export const cleanupError = (error: string, linesToRemove?: number = 1): string => {
    const [dirtyHeader, ...dirtyLines] = error.split('\n');

    const subtractLineNumber = (match: string, line: string): string => match.replace(line, safeSubtractFromString(line, linesToRemove));

    const header = dirtyHeader.replace(/\(([0-9]+):[0-9]+\)$/, subtractLineNumber);

    const lines = dirtyLines
        .map((line: string): string => line.replace(/^([0-9]+)\s*:/, subtractLineNumber))
        .filter((line: string): boolean => !line.match(/^\s*0+\s*:/));

    return [header, ...lines].join('\n');
};
