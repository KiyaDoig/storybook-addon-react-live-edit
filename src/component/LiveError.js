// @flow
import * as React from 'react';
import { withLive } from 'react-live';
import { className } from '../constants';
import { cleanupError } from '../returnToRender';

type LiveErrorProps = {
    live: {
        error?: ?string
    }
};

const LiveError = ({ live }: LiveErrorProps): ?React.Element<'div'> => live.error ? (
    <div className={className.error}>
        {cleanupError(live.error)}
    </div>
) : null;

export default withLive(LiveError);
