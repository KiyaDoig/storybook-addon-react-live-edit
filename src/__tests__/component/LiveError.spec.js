import * as React from 'react';
import { mount } from 'enzyme';
import { LiveProvider } from 'react-live';
import LiveError from '../../component/LiveError';
import returnToRender from '../../returnToRender';

describe('LiveError', () => {
    it('should not render with no error', () => {
        const wrapper = mount(
            <LiveProvider code="return <div />" transformCode={returnToRender} noInline>
                <LiveError />
            </LiveProvider>
        );

        expect(wrapper.find(LiveError).render()).toMatchSnapshot();
    });
    it('should render cleaned up error message for error', () => {
        const wrapper = mount(
            <LiveProvider code="re turn <div />" transformCode={returnToRender} noInline>
                <LiveError />
            </LiveProvider>
        );

        expect(wrapper.find(LiveError).render()).toMatchSnapshot();
    });
});
