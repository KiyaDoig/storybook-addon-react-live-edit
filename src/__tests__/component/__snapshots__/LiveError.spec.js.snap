// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveError should not render with no error 1`] = `null`;

exports[`LiveError should render cleaned up error message for error 1`] = `
<div
  class="live-edit-error"
>
  SyntaxError: Unexpected token (1:3)
1 : re turn &lt;div /&gt;
       ^
</div>
`;
